File: 2024/day-03/main2.ex
defmodule Parser do
  # Function to read the file content
  def read_file(filename) do
    case File.read(filename) do
      {:ok, content} -> content
      {:error, _} ->
        IO.puts("Error reading file: #{filename}")
        nil
    end
  end

  def process do
    # Set the file to read input from
    file = if length(System.argv()) > 0, do: List.first(System.argv()), else: "input.txt"

    # Read the input data from the file
    D = read_file(file)

    if D == nil, do: return

    part2 = process_data(D, 0, true)

    IO.puts(part2)
  end

  defp process_data("", part2, _enabled), do: part2
  defp process_data(D, part2, enabled) do
    {new_part2, new_enabled, rest} =
      case String.slice(D, 0, 4) do
        "do()" -> process_data(String.slice(D, 4..-1), part2, true)
        "don't()" -> process_data(String.slice(D, 7..-1), part2, false)
        "mul(" ->
          {new_part2, new_enabled, rest} = process_mul(D, part2, enabled)
          process_data(rest, new_part2, new_enabled)
        _ ->
          process_data(String.slice(D, 1..-1), part2, enabled)
      end

    new_part2
  end

  defp process_mul(D, part2, enabled) do
    # Find the closing parenthesis
    {mul_str, rest} = find_closing_paren(D, 4)

    # Extract numbers inside "mul(x, y)"
    case Regex.run(~r/(\d+),\s*(\d+)/, mul_str) do
      [_, x_str, y_str] ->
        x = String.to_integer(x_str)
        y = String.to_integer(y_str)

        # Make sure the closing parenthesis isn't followed by a number
        if rest != "" and !String.match?(String.slice(rest, 0, 1), ~r/\d/) do
          if enabled do
            part2 = part2 + x * y
          end
          IO.puts(mul_str)
        end

        {part2, enabled, rest}

      _ -> {part2, enabled, D}
    end
  end

  defp find_closing_paren(D, index) do
    # Find the closing parenthesis matching the opening "mul("
    {_, j} = Enum.reduce_while(index..String.length(D)-1, {0, index}, fn i, {depth, i_acc} ->
      if String.slice(D, i, i+1) == "(" do
        {:cont, {depth + 1, i+1}}
      else
        if String.slice(D, i, i+1) == ")" do
          if depth == 1 do
            {:halt, {i, String.slice(D, i+1..-1)}}
          else
            {:cont, {depth - 1, i+1}}
          end
        else
          {:cont, {depth, i+1}}
        end
      end
    end)
    {String.slice(D, index+1..j-1), j}
  end
end

# Run the program
Parser.process()

File: 2024/day-03/main2.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>

#define MAX_LINE_LENGTH 1024

// Function to read the file and return the content as a string
char* read_file(const char *filename) {
    FILE *file = fopen(filename, "r");
    if (file == NULL) {
        fprintf(stderr, "Error reading file: %s\n", filename);
        exit(1);
    }

    fseek(file, 0, SEEK_END);
    long file_size = ftell(file);
    fseek(file, 0, SEEK_SET);

    char *content = (char*)malloc(file_size + 1);
    if (content == NULL) {
        fprintf(stderr, "Memory allocation failed.\n");
        exit(1);
    }

    fread(content, 1, file_size, file);
    content[file_size] = '\0';

    fclose(file);
    return content;
}

// Function to process "mul(x, y)" operations
int process_mul(const char *data, int *index, int part2, int enabled) {
    int j = *index + 4;

    // Find the closing parenthesis
    while (data[j] != ')') {
        j++;
    }

    // Extract the substring for "mul(x, y)"
    char mul_str[j - (*index + 4) + 1];
    strncpy(mul_str, data + (*index + 4), j - (*index + 4));
    mul_str[j - (*index + 4)] = '\0';

    // Extract x and y using a simple manual split
    int x, y;
    if (sscanf(mul_str, "%d, %d", &x, &y) == 2) {
        // Make sure the closing parenthesis is not followed by a number
        if (!isdigit(data[j + 1])) {
            if (enabled) {
                part2 += x * y;
            }
            printf("%.*s\n", j + 1, data); // Print the "mul(x, y)" part
        }
    }

    *index = j + 1; // Update the index to move past the closing parenthesis
    return part2;
}

int process_data(const char *data, int part2, int enabled) {
    int i = 0;
    while (data[i] != '\0') {
        if (strncmp(&data[i], "do()", 4) == 0) {
            enabled = 1;
            i += 4;
        } else if (strncmp(&data[i], "don't()", 7) == 0) {
            enabled = 0;
            i += 7;
        } else if (strncmp(&data[i], "mul(", 4) == 0) {
            part2 = process_mul(data, &i, part2, enabled);
        } else {
            i++;
        }
    }
    return part2;
}

int main(int argc, char *argv[]) {
    // Set the file to read input from
    const char *file = (argc > 1) ? argv[1] : "inp
